assets/js/search.js

```diff
@@ -19,100 +19,82 @@ const $input = $('.sidebar-search input')
+        var endPos = matchPos[0] + matchPos[1] + nchars > item.doc.length
+          ? item.doc.length : matchPos[0] + matchPos[1] + nchars
+        var excerpt =
+          (startPos > 0 ? '...' : '') +
+          item.doc.slice(startPos, matchPos[0]) +
+          '<em>' +
+          item.doc.slice(matchPos[0], matchPos[0] + matchPos[1]) +
+          '</em> ' +
+          item.doc.slice(matchPos[0] + matchPos[1], endPos) +
+          (endPos < item.doc.length ? '...' : '')
+        excerpts.push(excerpt)
+      })
     }
-
-    return acc
-  }, acc || {})
-}
-
-function pushLevel (levels, searchEntity, name) {
-  if (searchEntity.length > 0) {
-    levels.push({name: name, results: searchEntity})
+  })
+  if (excerpts.length === 0) {
+    excerpts.push(item.doc.slice(0, nchars * 2))
   }
-}
-
-export function findIn (elements, matcher) {
-  return elements.map(function (element) {
-    let title = element.title
-    let titleMatch = title && title.match(matcher)
-    let result = {
-      id: element.id,
-      match: titleMatch ? highlight(titleMatch) : element.title
-    }
-    let hasMatch = !!titleMatch
-
-    if (element.nodeGroups) {
-      for (let {key, nodes} of element.nodeGroups) {
-        let matches = findNested(nodes, title, matcher, result[key])
-        if (Object.keys(matches).length > 0) {
-          hasMatch = true
-          if (key === 'types' || key === 'callbacks') {
-            result[key] = matches
-          } else {
-            result.functions = matches
-          }
-        }
-      }
-    }
-
-    if (hasMatch) {
-      for (let key in result) {
-        if (key !== 'id' && key !== 'match') {
-          result[key] = Object.values(result[key]).sort((a, b) => a.id.localeCompare(b.id))
-        }
-      }
-
-      return result
-    }
-  }).filter(cleaner)
+  return excerpts.slice(0, 1)
 }
 
 export function search (value) {
-  var nodes = sidebarNodes
-
   if (value.replace(/\s/, '') !== '') {
-    var safeVal = new RegExp(helpers.escapeText(value), 'i')
-    var levels = []
-
-    var modules = findIn(nodes.modules, safeVal)
-    var exceptions = findIn(nodes.exceptions, safeVal)
-    var tasks = findIn(nodes.tasks, safeVal)
-
-    // add to the results
-    pushLevel(levels, modules, 'Modules')
-    pushLevel(levels, exceptions, 'Exceptions')
-    pushLevel(levels, tasks, 'Mix Tasks')
-
-    var results = resultsTemplate({
+    var idx = getIndex()
+    var results = idx.search(value)
+    results = fillResults(results, value)
+    var resultsHtml = resultsTemplate({
       value: value,
-      levels: levels,
-      empty: levels.length === 0
+      results: results,
+      empty: results.length === 0
     })
 
     $input.val(value)
-    $search.html(results)
+    $search.html(resultsHtml)
   }
 }
+
+function getIndex () {
+  if (sessionStorage.getItem('idx') == null) {
+    createIndex()
+  }
+  return lunr.Index.load(JSON.parse(sessionStorage.getItem('idx')))
```

Comment:
We also need to check what happens if the data is too large to fit the session storage. I assume it will fail to store and getItem will then fail? so it may be better to write this feature in a way that it works even if sessionStorage fails, something like:

```js
var index, stored = sessionStorage.getItem('idx');
if(stored == null) {
  index = createIndex()
  sessionStorage.setItem('idx', JSON.stringify(idx))
} else {
  index = JSON.parse(stored)
}
return lunr.Index.load(index)
```